## Chapter 1: Understanding the Data

  ### Business Context
  Every e-commerce business faces the same core challenge — understanding
  their customers. Who are the loyal ones? Who is about to leave? Who spends
  the most? Without answers to these questions, businesses make decisions
  based on gut feeling rather than data.

  In this project we are building a Customer Intelligence Platform that answers
  these questions using real transaction data from a UK-based online retailer
  covering December 2010 to December 2011.

  ### About the Dataset
  The dataset contains 541,909 transactions across 38 countries with 8 columns:
  - InvoiceNo: Unique identifier for each transaction
  - StockCode: Product identifier
  - Description: Product name
  - Quantity: Number of units purchased
  - InvoiceDate: Date and time of purchase
  - UnitPrice: Price per unit in GBP
  - CustomerID: Unique identifier for each customer
  - Country: Country where the customer is based

  ### What are we trying to find?
  Before building any model we need to understand:
  1. How clean is the data?
  2. Are there any anomalies or data quality issues?
  3. What does a typical customer look like?

In [3]:
import pandas as pd

In [4]:
df = pd.read_excel('/Users/purvamugdiya/Desktop/customer-intelligence-platform/data/Online Retail.xlsx')

Data Exploration

In [6]:
#Basic exploration
print("\nShape of the dataset:", df.shape)
print("\nColumns in the dataset:", df.columns.to_list())
print("\nFirst 5 rows of the dataset:")
df.head()


Shape of the dataset: (541909, 8)

Columns in the dataset: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']

First 5 rows of the dataset:


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [5]:
print("\nData types of each column:")
print(df.dtypes)
print("Missing values in each column:\n", df.isnull().sum())
print("\nUnique values in 'Country' column:", df['Country'].nunique())
print("\nTop 5 countries by number of transactions:")
print(df['Country'].value_counts().head())  
print("\nSummary statistics for numerical columns:")
df.describe()


Data types of each column:
InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID            float64
Country                object
dtype: object
Missing values in each column:
 InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

Unique values in 'Country' column: 38

Top 5 countries by number of transactions:
Country
United Kingdom    495478
Germany             9495
France              8557
EIRE                8196
Spain               2533
Name: count, dtype: int64

Summary statistics for numerical columns:


,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386048,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


### What We Found — Key Observations

  **1. Scale of the data**
  541,909 transactions is substantial real-world data. This is not a toy
  dataset — it represents an entire year of actual business operations.

  **2. The missing CustomerID problem**
  135,080 rows (25% of data) have no CustomerID. These are guest purchases —
  customers who bought without creating an account. Since all our models need
  to track individual customer behavior over time, we cannot use these rows.
  This is a common real-world data quality challenge.

  **3. Negative quantities and prices**
  The minimum Quantity is -80,995 and minimum UnitPrice is -11,062. These
  are not errors — they represent product returns and order cancellations.
  A business needs to track returns, but for our customer behavior models
  these would distort our analysis. We will remove them.

  **4. Cancelled orders**
  Some InvoiceNo values start with 'C' — these are explicitly cancelled orders.
  Including them would make customers appear to have purchased more than they
  actually did.

  ### So What?
  Raw data is never ready for modeling. Understanding these issues upfront
  means our models will be built on accurate, trustworthy data — which
  directly determines how reliable our predictions will be.